manual/src/tutorials/parallelism.etex

```diff
@@ -0,0 +1,732 @@
+  task' ();
+  Domain.join d
+
+let _ = main ()
+\end{verbatim}
+
+\begin{verbatim}
+$ ocamlopt -I +threads unix.cmxa threads.cmxa -o dom_thr.exe dom_thr.ml
+$ ./dom_thr.exe
+Thread 1 running on domain 1 saw initial write
+Thread 0 running on domain 0 saw the write by thread 1
+Thread 2 running on domain 1 saw the write by thread 0
+Thread 3 running on domain 0 saw the write by thread 2
+\end{verbatim}
+
+This program uses a shared reference cell protected by a mutex to communicate
+between the different systhreads running on two different domains. The
+systhread identifiers uniquely identify systhreads in the program. The initial
+domain gets the domain id and the thread id as 0. The newly spawned domain gets
+domain id as 1.
+
+\section{s:par_atomics}{Atomics}
+
+Mutex, condition variables and semaphores are used to implement blocking
+synchronisation between domains. For non-blocking synchronisation, OCaml
+provides \stdmoduleref{Atomic} variables. As the name suggests, non-blocking
+synchronisation does not provide mechanisms for suspending and waking up
+domains. On the other hand, primitives used in non-blocking synchronisation are
+often compiled to atomic read-modify-write primitives that the hardware
+provides. As an example, the following program increments a non-atomic counter
+and an atomic counter in parallel.
+
+\begin{caml_example*}{verbatim}
+(* incr.ml *)
+let twice_in_parallel f =
+  let d1 = Domain.spawn f in
+  let d2 = Domain.spawn f in
+  ignore @@ Domain.join d1;
+  ignore @@ Domain.join d2
+
+let plain_ref n =
+  let r = ref 0 in
+  let f () = for _i=1 to n do incr r done in
+  twice_in_parallel f;
+  Printf.printf "Non-atomic ref count: %d\n" !r
+
+let atomic_ref n =
+  let r = Atomic.make 0 in
+  let f () = for _i=1 to n do Atomic.incr r done in
+  twice_in_parallel f;
+  Printf.printf "Atomic ref count: %d\n" (Atomic.get r)
+
+let main () =
+  let n = try int_of_string Sys.argv.(1) with _ -> 1 in
+  plain_ref n;
+  atomic_ref n
+
+let _ = main ()
+\end{caml_example*}
+
+\begin{verbatim}
+$ ocamlopt -o incr.exe incr.ml
+$ ./incr.exe 1_000_000
+Non-atomic ref count: 1187193
+Atomic ref count: 2000000
+\end{verbatim}
+
+Observe that the result from using the non-atomic counter is lower than what
+one would naively expect. This is because the non-atomic "incr" function is
+equivalent to:
+
+\begin{caml_example*}{verbatim}
+let incr r =
+  let curr = !r in
+	r := curr + 1
+\end{caml_example*}
+
+Observe that the load and the store are two separate operations, and the
+increment operation as a whole is not performed atomically. When two domains
+execute this code in parallel, both of them may read the same value of the
+counter "curr" and update it to "curr + 1". Hence, instead of two increments,
+the effect will be that of a single increment. On the other hand, the atomic
+counter performs the load and the store atomically with the help of hardware
+support for atomicity. The atomic counter returns the expected result.
```

Comment:
> We could add here an additional explanation here to emphasize that this is expected because this piece of code contains a data race.

This is what I had originally :-) And @fpottier rightly pointed out that there is the simpler issue of non-atomicity. I'm in two minds about discussing the memory model issues especially when the protocol itself is non-atomic. The readers may not have sufficient understanding of the relaxed memory model issues to appreciate the observations that we may make.

Reply:
I'm choosing to leave the current text as is here.